File: src/main/scala/com/vc/hard/L295.scala
package com.vc.hard

class MedianFinder() {

  /** initialize your data structure here. */
  import scala.collection.mutable
  object PQOrdering extends Ordering[Int] {
    def compare(x: Int, y: Int): Int = y.compareTo(x)
  }

  val maxHeap = new mutable.PriorityQueue[Int]()(PQOrdering)
  val minHeap = new mutable.PriorityQueue[Int]()
  def addNum(num: Int) {
    minHeap += num
    maxHeap += minHeap.dequeue
    if(minHeap.size < maxHeap.size) {
      minHeap += maxHeap.dequeue
    }
  }

  def findMedian(): Double = {
    if(maxHeap.size < minHeap.size) minHeap.head
    else (maxHeap.head + minHeap.head) / 2.0
  }
}

/**
  * Your MedianFinder object will be instantiated and called as such:
  * var obj = new MedianFinder()
  * obj.addNum(num)
  * var param_2 = obj.findMedian()
  */

File: src/main/scala/com/vc/hard/L295.java
package com.vc.hard;

import java.util.*;

class MedianFinder {

    /** initialize your data structure here. */
    PriorityQueue<Integer> left = new PriorityQueue<>((a, b) -> b - a);
    PriorityQueue<Integer> right = new PriorityQueue<>((a, b) -> a - b);

    public void addNum(int num) {
        if(left.size() <= right.size()) {
            right.offer(num);
            left.offer(right.poll());
        }
        else {
            left.offer(num);
            right.offer(left.poll());
        }
    }

    public double findMedian() {
        if(left.size() == right.size()) return ((double)left.peek() + (double)right.peek()) / 2.0;
        else return (double)left.peek();
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */
